### PR TITLE
Add BIP32 Support

### PR DESCRIPTION
**Pull request by SachinMeier**
- Extended Keys
- Derivation Paths
- Seed -> Master Extended Key

### CODE REVIEW

File: lib/extendedkey.ex [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+    @moduledoc """
+    Contains a list of integers (or the :any atom) representing a bip32 derivation path. 
+    The :any atom represents a wildcard in the derivation path. DerivationPath structs can
+    be used by ExtendedKey.derive_extended_key to derive a child key at the given path.
+    """
+
+    @min_non_hardened_child_num 0
+    # 2^31 - 1
+    @max_non_hardened_child_num 0x80000000 - 1
+    @min_hardened_child_num 0x80000000
```
Comment (philipglazman) [inline lib/extendedkey.ex:20]: We could simplify these constants to just include `min_hardened_child_num` and `max_hardened_child_num`, the other ones do not seem to be useful

File: lib/extendedkey.ex [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+        [""] -> []
+        ["m" | rest] -> tstring_to_path(rest)
+        ["*" | rest] -> [:any | tstring_to_path(rest)]
+        [i | rest] -> [str_to_level(i) | tstring_to_path(rest)]
+      end
+    end
+
+    defp str_to_level(level) do
+      {num, is_hardened} =
+        case String.split(level, ["'", "h"]) do
```
[the remark this answers is not in the document]
Comment (SachinMeier) [reply lib/extendedkey.ex:98]: Yeah, same. I prefer apostrophe, but this allows us to parse both, so i figured flexibility was good. 

We can parse flexibly but serialize using apostrophe.

File: lib/extendedkey.ex [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+    defstruct [:child_nums]
+
+    def max_non_hardened_child_num(), do: @max_non_hardened_child_num
+    def max_hardened_child_num(), do: @max_hardened_child_num
+
+    @spec path_to_string(t()) :: String.t()
+    def path_to_string(%__MODULE__{child_nums: path}), do: tpath_to_string(path, "")
+
+    defp tpath_to_string([], path_acc), do: path_acc
+    defp tpath_to_string(["m" | rest], path_acc), do: tpath_to_string(rest, "m/" <> path_acc)
```
[the remark this answers is not in the document]
Comment (SachinMeier) [reply lib/extendedkey.ex:47]: I think we can take this line out. When parsing strings, it discards the "m"s anyway.

File: test/extendedkey_test.exs [hunk 4]
```diff
@@ -0,0 +1,764 @@
...
+    xprv_m:
+      "xprv9s21ZrQH143K25QhxbucbDDuQ4naNntJRi4KUfWT7xo4EKsHt2QJDu7KXp1A3u7Bi1j8ph3EGsZ9Xvz9dGuVrtHHs7pXeTzjuxBrCmmhgC6",
+    xpub_m_0h:
+      "xpub68NZiKmJWnxxS6aaHmn81bvJeTESw724CRDs6HbuccFQN9Ku14VQrADWgqbhhTHBaohPX4CjNLf9fq9MYo6oDaPPLPxSb7gwQN3ih19Zm4Y",
+    xprv_m_0h:
+      "xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L"
+  }
+
+  @bip49_test_case %{
+    masterseedWords:
```
[the remark this answers is not in the document]
Comment (SachinMeier) [reply test/extendedkey_test.exs:96]: ill comment it out.

File: lib/extendedkey.ex [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+  # Single Sig
+  # xpub
+  @xpub_pfx <<0x04, 0x88, 0xB2, 0x1E>>
+  # xprv
+  @xprv_pfx <<0x04, 0x88, 0xAD, 0xE4>>
+  # tpub
+  @tpub_pfx <<0x04, 0x35, 0x87, 0xCF>>
+  # tprv
+  @tprv_pfx <<0x04, 0x35, 0x83, 0x94>>
+  # ypub
```
[the remark this answers is not in the document]
Comment (SachinMeier) [reply lib/extendedkey.ex:162]: ok. Should we remove this support then? Wallets like Wasabi and BlueWallet will follow bip44,49,84 and derive different wallets based on the prefixes.

File: lib/extendedkey.ex (hunk 1 above)
Comment (SachinMeier) [reply lib/extendedkey.ex:20]: Kafai and I discussed this on his PR. he thought it would provide more clarity

File: test/extendedkey_test.exs [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+    "xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet",
+    # invalid len
+    "xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1E",
+    # invalid len (with valid checksum)
+    "Deb7pNXSbX7qSvc2eMjkNYTrggh4pBgYa2QMFjEjj6hUy1iHp71d1gf3ue7Ni7X8Pkhcy13aakoEYSMS1DFiMaXB4qYhN33dh3oqD9n1YWSxVk"
+  ]
+
+  @bip32_test_case_1 %{
+    # test vectors from bip32: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#test-vectors
+    seed: "000102030405060708090a0b0c0d0e0f",
```
Comment (philipglazman) [inline test/extendedkey_test.exs:28]: In the test cases, we have seed, but we do not test derivation from the seed enough.

The test coverage for parsing looks good. Would be nice to test derivation from seed for all test cases.

File: test/extendedkey_test.exs [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+    xprv_m_0_2147483647h_1_2147483646h:
+      "xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc",
+    xpub_m_0_2147483647h_1_2147483646h_2:
+      "xpub6FnCn6nSzZAw5Tw7cgR9bi15UV96gLZhjDstkXXxvCLsUXBGXPdSnLFbdpq8p9HmGsApME5hQTZ3emM2rnY5agb9rXpVGyy3bdW6EEgAtqt",
+    xprv_m_0_2147483647h_1_2147483646h_2:
+      "xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j"
+  }
+
+  @bip32_test_case_3 %{
+    # Test for retention of leading Zeros
```
Comment (philipglazman) [inline test/extendedkey_test.exs:111]: add seed for test case 3

`4b381541583be4423346c643850da4b320e46a87ae3d2a4e6da11eba819cd4acba45d239319ac14f863b8d5ab5a0d0c64d2e8a1e7d1457df2e5a3c51c73235be`

File: lib/extendedkey.ex [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+    @type t :: %__MODULE__{
+            child_nums: list(child_num)
+          }
+
+    @enforce_keys [
+      :child_nums
+    ]
+    defstruct [:child_nums]
+
+    def max_non_hardened_child_num(), do: @max_non_hardened_child_num
```
[the remark this answers is not in the document]
Comment (SachinMeier) [reply lib/extendedkey.ex:40]: agreed

File: test/extendedkey_test.exs (hunk 7 above)
Comment (SachinMeier) [reply test/extendedkey_test.exs:111]: i concur

File: test/extendedkey_test.exs (hunk 6 above)
Comment (SachinMeier) [reply test/extendedkey_test.exs:28]: I added testing for the seed -> master privkey. What did you mean "for all test cases"?

Comment (philipglazman) [reply test/extendedkey_test.exs:28]: Using the bip32 test cases as a crutch, assert that the expected child matches the derived child.

For example, make sure our derivation works from seed to the deepest child provided in the test cases.
seed -> master privkey -> xpub_m_0h_1_2h -> xpub_m_0h_1_2h_2_1000000000

Comment (SachinMeier) [reply test/extendedkey_test.exs:28]: yes, makes sense. ill add now

**Review by philipglazman — APPROVED** [review]

**Pull request merged**